### PR TITLE
[Octo-Bot] Adding new features to healthcare-web-bot

### PR DESCRIPTION
**Pull request by joelczk**

### CODE REVIEW

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testpermissionslogs.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+
+from testUtil.testScrapeAdmin.testadmin import viewAccountPages
+
+paginationXpath = "/html/body/form/div[4]/div[4]/div/div/table/tbody/tr[22]/td/table/tbody/tr/td["
+
+""" Saves all the permission logs of the website
+
+savePermissionLogs obtains all the perrmission logs for one page
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testpermissionslogs.py:15]:
```suggestion
savePermissionLogs obtains all the permission logs for one page
```

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testrecordlogs.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+from bs4 import BeautifulSoup
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+
+from testUtil.testDriver.testrole import isAdmin
+from testUtil.testScrapeAdmin.testadmin import viewAccountPages
+
+from prettytable import PrettyTable
+
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testrecordlogs.py:12]: missing function documentation!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/testviewtherapist.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                if line == "View":
+                    continue
+                else:
+                    info.append(line)
+            if len(info) == 0:
+                continue
+            else:
+                f.write("Therapist Name: " + str(info[0]) + "\n")
+                f.write("     -Title: " + str(info[1]) + "\n")
+                f.write("     -Department: " + str(info[2]) + "\n")
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/testviewtherapist.py:51]: Python best-practice, newline at the last line :)
```suggestion
                f.write("     -Department: " + str(info[2]) + "\n")
                
```

File: Octo-Bot/healthcare-web-bot/test/testmain.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+from testUtil.testScrapeTherapist.viewPatients import iteratePage, loginTherapist
+from testUtil.testScrapePatient.viewDiagnosis import getDiagnosisInformation
+
+def getDriver(username, password, url):
+    
+    '''
+    Obtains the driver to be used to start the firefox instance
+
+    Arguments:
+        username(str) : Username of credentials that the driver is to be logging in to
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testmain.py:35]:
```suggestion
        username(str) : Username of credentials that the driver is logging in to
```

File: Octo-Bot/healthcare-web-bot/util/Driver/adminparser.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            number += 1
+            continue
+        else:
+            return False
+    return True
+
+def genAdminFiles(botNumbers):
+    
+    '''
+    Generate csv files for different usernames-passwords according to the number of botd
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/util/Driver/adminparser.py:57]: typo?
```suggestion
    Generate csv files for different usernames-passwords according to the number of bots
```

File: Octo-Bot/healthcare-web-bot/util/Driver/daemon.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    Returns:
+        None
+    '''
+    while(True):
+        thread = threading.Thread(target = daemonActions, args = ())
+        thread.daemon = True
+        thread.start()
+        time.sleep(1)
+        thread.join()
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/util/Driver/daemon.py:35]: Python best-practice, newline at the last line :)
```suggestion
        thread.join()
        
```

**Comment by ariscahyadi** [comment]
When the docker run, there is error `Error in csv file. Please check again!`

Please update the documentation to update the CSV file with correct credential.

**Comment by ariscahyadi** [comment]
I received this error when I tried locally, I think due to host unreachable.

```
$ sudo docker run healthcare-web-bot:latest -r admin
Preparing to scrap website...
37
Reading from //config/admin/adminLogin15.csv
Username used: S0000001A
Password used: <>
File Number used: 15
Error in csv file. Please check again!
```

File: Octo-Bot/healthcare-web-bot/Docs/README.md [hunk 7]
```diff
@@ -1,4 +1,6 @@
 # Table of content
-1. [Configuration](configuration.md)
-2. [Frequently Asked Questions (FAQ)](FAQ(Frequently%20Asked%20Questions).md)
-3. [Developing the bot](Developing%20the%20Bot.md)
+1. [Set-up](setup.md)
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/Docs/README.md]: Resolved!

File: Octo-Bot/healthcare-web-bot/Docs/setup.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+joelczk@OctoBot:~$ sudo apt-get update
+joelczk@OctoBot:~$ sudo apt-get install docker-ce docker-ce-cli containerd.io
+joelczk@OctoBot:~$ sudo docker run hello-world
+```
+
+For other distributions, visit [here](https://docs.docker.com/engine/install/)
+
+You can then build the docker image using the provided script like this:
+```console
+joelczk@OctoBot/Octo-Bot/<bot-name>:~$./build.sh
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/Docs/setup.md:29]: Resolved!

File: Octo-Bot/healthcare-web-bot/Docs/Usage.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+Arguments for program
+
+optional arguments:
+  -h, --help            show this help message and exit
+  -t time               Time to sleep between crawling of website links
+  -r role               Role that the bot should login as
+  -b bot                Number of credentials for the choosen role
+  -u username           Username to be used for login
+  -p password           Password to be used for login
+  -m mode of execution  Mode for bot to be executed
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/Docs/Usage.md]: Resolved!

File: Octo-Bot/healthcare-web-bot/Docs/Usage.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+## Crawl the website with multiple bots in parallel
+Specify the number of parallel bots that are to be run at the same time. The default value is set to 1 if this field is not specified.
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -b 5
+```
+
+## Crawl the bot with specified username and password
+Specify the username and password to login and crawl the website. The bot will login from the csv list of usernames and passwords if this is not specified.
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -u <usernam> -p <password>
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/Docs/Usage.md]: Resolved!

File: Octo-Bot/healthcare-web-bot/main.py [hunk 11]
```diff
@@ -35,57 +51,223 @@ def getDriver(username, password, url):
...
+    else:
+        fileNumber = "No files are used"
+        username = checkUsername
+        password = checkPassword
+    print("Username used: " + str(username))
+    print("Password used: " + str(password))
+    print("File Number used: " + str(fileNumber))
     driver = getDriver(username, password, url)
-    func(driver)
+    if driver.current_url == "https://10.10.0.112/Therapist/Dashboard":
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/main.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/test.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    testPatientCSV tests if the number of records in the patient.csv files tally
+    testTherapistCSV tests if the number of records in the therapist.csv files tally
+    testDaemon tests if the daemon mode is running properly
+    """
+    def setUp(self):
+        warnings.simplefilter('ignore',category = DeprecationWarning)
+        warnings.simplefilter('ignore', category = ResourceWarning)
+        
+    def testUrl(self):
+        url = "https://10.10.0.112"
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/test.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/test.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        self.assertEqual(number, 102)
+    
+    def testTherapistCSV(self):
+        number = therapistGetNumberRecords()
+        self.assertEqual(number, 123)
+    
+    def testDriver(self):
+        username = "S1234567B"
+        password = "easyP@ssw0rd"
+        url = "http://10.10.0.112"
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/test.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/adminparser.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
+import codecs
+import csv
+import json
+import os
+import random
+import sys
+
+directory = str(os.getcwd())
+
+final_data = {"url": "http://10.10.0.112"}
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/adminparser.py:10]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/test.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        sys.stdout = out
+        daemonMode
+        sys.stdout = sys.__stdout__
+        if "Finishing of thread" in out.getvalue():
+            isDaemonMode = False
+        self.assertTrue(isDaemonMode)
+
+
+if __name__ == '__main__':
+    unittest.main()
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/test.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/daemon.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+    Returns:
+        None
+    '''
+    while(True):
+        thread = threading.Thread(target = daemonActions, args = ())
+        thread.daemon = True
+        thread.start()
+        time.sleep(1)
+        thread.join()
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/daemon.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/parser.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
+import codecs
+import csv
+import json
+import os
+import random
+import sys
+
+directory = str(os.getcwd())
+
+final_data = {"url": "http://10.10.0.112"}
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/parser.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/patientparser.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+    Returns:
+        None
+    '''
+    fileDirectory = directory + "/config/patient/patientLogin" + str(fileNumber) + ".csv"
+    writeFile = open(fileDirectory,mode = 'a', newline = '')
+    writer = csv.writer(writeFile, delimiter = ',')
+    writeBack = []
+    writeBack.append(username)
+    writeBack.append(password)
+    writer.writerow(writeBack)
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/patientparser.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/testdaemon.py [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+    while(True):
+        time.sleep(2)
+
+def daemonMode():
+    while(True):
+        thread = threading.Thread(target = daemonActions, args = ())
+        thread.daemon = True
+        thread.start()
+        time.sleep(1)
+        thread.join()
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/testdaemon.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/testtherapistparser.py [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    return credentials
+
+def therapistWriteBack(username, password, fileNumber):
+    fileDirectory = directory + "/config/therapist/therapistLogin" + str(fileNumber) + ".csv"
+    writeFile = open(fileDirectory,mode = 'a', newline = '')
+    writer = csv.writer(writeFile, delimiter = ',')
+    writeBack = []
+    writeBack.append(username)
+    writeBack.append(password)
+    writer.writerow(writeBack)
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/testtherapistparser.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/therapistparser.py [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+    Returns:
+        None
+    '''
+    fileDirectory = directory + "/config/therapist/therapistLogin" + str(fileNumber) + ".csv"
+    writeFile = open(fileDirectory,mode = 'a', newline = '')
+    writer = csv.writer(writeFile, delimiter = ',')
+    writeBack = []
+    writeBack.append(username)
+    writeBack.append(password)
+    writer.writerow(writeBack)
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testDriver/therapistparser.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testpermissionslogs.py (hunk 1 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testpermissionslogs.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testrecordlogs.py (hunk 2 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeAdmin/testrecordlogs.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/testviewtherapist.py (hunk 3 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/testviewtherapist.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testmain.py (hunk 4 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testmain.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/util/Driver/adminparser.py (hunk 5 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/util/Driver/adminparser.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/util/Driver/daemon.py (hunk 6 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/util/Driver/daemon.py]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/newrecord.py [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+def getWeight():
+
+    '''
+    Obtain weight information to be injected in
+
+    Arguments: 
+        None
+    
+    Returns:
+        Dictionary object of weight readigs
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/newrecord.py:55]:
```suggestion
        Dictionary object of weight readings
```

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/newrecord.py [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+def getBloodPressure():
+
+    '''
+    Obtains blood pressure information to be injected
+
+    Arguments:
+        None
+    
+    Returns:
+        Dictionary object of blodd pressure readings
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/newrecord.py:121]:
```suggestion
        Dictionary object of blood pressure readings
```

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py [hunk 24]
```diff
@@ -0,0 +1,277 @@
...
+        except Exception as e:
+            print(str(e))
+            print("Path creation failed!")  
+    else:
+        return
+
+def numberOfEnteries(driver):
+
+    '''
+    Obtains the number of enteries of therapist information
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py:35]:
```suggestion
    Obtains the number of entries of therapist information
```

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py [hunk 25]
```diff
@@ -0,0 +1,277 @@
...
+
+    Arguments:
+        driver(obj): firefox webdriver instance in python 
+    
+    Returns:
+        Total number of enteries for therapist information
```
Comment (ariscahyadi) [inline Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py:41]:
```suggestion
        Total number of entries for therapist information
```

File: Octo-Bot/healthcare-web-bot/README.md [hunk 26]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+optional arguments:
+  -h, --help   show this help message and exit
+  -t time      Time to sleep between crawling of website links
+  -r role      Role that the bot should login as
+  -b bot       Number of bots for the choosen role
+  -u username  Username to be used for login
+  -p password  Password to be used for login
+```
+
+# Examples
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:57]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 27]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+
+## Crawl with waiting time in between website links
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:59]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 28]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+This allows the bot to crawl website links with a specified sleep time between crawling. The default value is set to 0, if this is not specified
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -t 20
+```
+
+## Crawl the website with a specified role
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:65]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 29]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+This allows the bot to crawl the website using an intended role. The roles available are "admin", "patient", "researcher" and "therapist". The bot will take on the default role of an admin if this is not specified.
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -r admin
+```
+
+## Crawl the website with multiple bots in parallel
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:71]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 30]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+Specify the number of parallel bots that are to be run at the same time. The default value is set to 1 if this field is not specified.
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -b 5
+```
+
+## Crawl the bot with specified username and password
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:77]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 31]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+Specify the username and password to login and crawl the website. The bot will login from the csv list of usernames and passwords if this is not specified.
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -u <usernam> -p <password>
+```
+## Execute the bot in a daemon mode
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:82]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 32]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
 joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo chmod -R <directory>
 joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ cd <directory>/data
 ```
+
+## Usage
+Here is the help page for the healthcare-web-bot program:
+```console
+joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ docker run healthcare-web-bot -h
+usage: main.py [-h] [-t time] [-r role] [-b bot] [-u username] [-p password]
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:45]: Resolved!

File: Octo-Bot/healthcare-web-bot/README.md [hunk 33]
```diff
@@ -37,4 +37,52 @@ joelczk@OctoBot:~/OctoBot/Octo-Bot/<bot-name>$ sudo docker cp <container name>:/
...
+Arguments for program
+
+optional arguments:
+  -h, --help   show this help message and exit
+  -t time      Time to sleep between crawling of website links
+  -r role      Role that the bot should login as
+  -b bot       Number of bots for the choosen role
+  -u username  Username to be used for login
+  -p password  Password to be used for login
+```
```
[the remark this answers is not in the document]
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/README.md:55]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/newrecord.py (hunk 23 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testScrapePatient/newrecord.py:121]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py (hunk 24 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py:35]: Resolved!

File: Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py (hunk 25 above)
Comment (joelczk) [reply Octo-Bot/healthcare-web-bot/test/testUtil/testScrapeTherapist/viewPatients.py:41]: Resolved!

**Pull request merged**